### PR TITLE
runelite-client: Add custom notifications

### LINKED ISSUE #2060: Windows 10 custom notifications
In Win10 you can't click in the red area. 
unbenannt

### PR DESCRIPTION
**Pull request by Abextm**
This is enabled by default on windows, because windows' notifications are broken for a large amount of users.

Fixes #2060

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java [hunk 1]
```diff
@@ -125,6 +125,8 @@
 	private final EventBus eventBus;
 	private final KeyManager keyManager;
 	private Applet client;
+
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: This effectively destroys the purpose of ClientUI, I would rather not expose the frame.

File: runelite-client/src/main/java/net/runelite/client/Notifier.java [hunk 2]
```diff
@@ -135,6 +158,112 @@ private void sendNotification(
...
+		final String subtitle)
+	{
+		final int width = 360;
+		final int height = 100;
+		final int padding = 50;
+
+		JFrame parentWindow = clientUI.get().getFrame();
+		Rectangle display = parentWindow.getGraphicsConfiguration().getBounds();
+
+		JFrame window = new JFrame();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/Notifier.java]: This should be probably custom component in some separate class

File: runelite-client/src/main/java/net/runelite/client/Notifier.java [hunk 3]
```diff
@@ -135,6 +158,112 @@ private void sendNotification(
...
+			height
+		);
+		((JPanel) window.getContentPane()).putClientProperty(SubstanceSynapse.COLORIZATION_FACTOR, 1.0);
+
+		window.addMouseListener(new MouseAdapter()
+		{
+			@Override
+			public void mousePressed(MouseEvent e)
+			{
+				window.setVisible(false);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/Notifier.java]: You should probably cancel the timer here too

File: runelite-client/src/main/java/net/runelite/client/Notifier.java [hunk 4]
```diff
@@ -135,6 +158,112 @@ private void sendNotification(
...
+		{
+			window.setVisible(false);
+			window.dispose();
+		}).start();
+
+		float[] hsbs = new float[3];
+		Color stock = window.getContentPane().getBackground();
+		Color.RGBtoHSB(stock.getRed(), stock.getGreen(), stock.getBlue(), hsbs);
+		final long start = System.currentTimeMillis();
+		Timer anim = new Timer(20, e ->
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/Notifier.java]: I thought creating threads like this (timer is creating thread i guess) is generally disliked

File: runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java (hunk 1 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: How should I not expose the frame?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Create the notification from inside the ClientUI.

File: runelite-client/src/main/java/net/runelite/client/Notifier.java (hunk 4 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: There is no other way that I know of to do this.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: Would be nice to extend current scheduler implementation to work outside of plugins probably, I needed that too before

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: Even if we did that that wouldn't help because I need to cancel

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: Its very unnoticeable if I don't animate it

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: Also I don't think this creates a thread, because it runs on the EDT

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java [hunk 5]
```diff
@@ -116,18 +116,29 @@ default boolean sendNotificationsWhenFocused()
...
 
+	@ConfigItem(
+		keyName = "customNotifications",
+		name = "Use non-native notifications",
+		description = "Use RuneLite styled notifications",
+		position = 14
+	)
+	default boolean customNotifications()
+	{
+		return System.getProperty("os.name").toLowerCase().contains("win");
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: I would make this disabled by default, this is problem only on Windows 10

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Windows 8+ has this problem

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Still. Also, not everyone complained about this, I would just make this opt-in every time, as native system functionality should be always preferred.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Native functionally is FUBAR for most users.

Comment (Jalopyy) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: The problem has to do with W10 toasts (the shitty replacement for balloon notifications) have dead zones. You just have to either make a regedit or download a program to revert them and you won't have this issue.

File: runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java [hunk 6]
```diff
@@ -760,4 +769,114 @@ private void saveClientBoundsConfig()
...
+		final TrayIcon.MessageType type,
+		final String subtitle)
+	{
+		final int width = 360;
+		final int height = 100;
+		final int padding = 50;
+
+		Rectangle display = frame.getGraphicsConfiguration().getBounds();
+
+		JFrame window = new JFrame();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: I still think this should be separate class. Also, probably using JDialog instead of JFrame would make more sense here.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Using JDialog is incorrect because a modal notification would be terrible.
As I said on discord, making this a separate class would just be a ctor with this method, because extending JFrame would expose the possibly invalid JFrame.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Yes exaclty, just like the SwingUtil class works. I mainly disagree with the amount of code in this method as I said earlier. Even having new Notification(frame) would be still better than having this lying around either in Notifier or ClientUI

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Also this will now expose 2 public apis for creating notifications, that otherwise would not need to exist.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Yes. That sounds correct. Look at the new notification as new reusable component, having it separate is correct. And, the method in ClientUI as factory for displaying this component for anyone who would want to d it.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: So its okay if we expose a notification that can be misused, but not okay if we expose the main jframe that users won't notice if its misused?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: Yes. The users not notificing if the jframe is misused is debatable, because you still can change all it's properties from outside when it will be exposed. Exposing the Notification instead of entire JFrame will result in less dumb PRs.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java]: I still disagree, but it has been done.

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.Timer;
+import lombok.extern.slf4j.Slf4j;
+import org.pushingpixels.substance.internal.SubstanceSynapse;
+
+@Slf4j
+class CustomNotification
+{
+	private static final int width = 360;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: UPPER_CASE

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+			public void mousePressed(MouseEvent e)
+			{
+				window.setVisible(false);
+				window.dispose();
+				anim.stop();
+				closeTimer.stop();
+			}
+		});
+
+		if (runeliteIcon == null)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: To not have everything in this method, initializing this icon in static constructor might make sense

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+			}
+		}
+
+		GroupLayout layout = new GroupLayout(window.getContentPane());
+		window.setLayout(layout);
+
+		JLabel iconLabel = new JLabel(new ImageIcon(runeliteIcon));
+		JLabel titleLabel = new JLabel(title);
+		// Not using the runescape fonts, because they need to be scaled, and look awful
+		titleLabel.setFont(new Font("sansserif", Font.BOLD, 14));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: This font name can probably be constant

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	private static final int height = 100;
+	private static final int padding = 50;
+
+	private static BufferedImage runeliteIcon;
+
+	static void sendCustomNotification(
+		final String title,
+		final String message,
+		final TrayIcon.MessageType type,
+		final String subtitle,
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: Subtitle and type is not even used in this method

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+		final JFrame runeliteWindow)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: This should be probably simply `parent`

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java (hunk 8 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: I lazy loaded it as it won't be used in many cases

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java (hunk 10 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: It is to conform to the notifier api, in case we ever use those parameters in the future. The other notifiers often don't read subtitle either.

File: runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java (hunk 8 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/ui/CustomNotification.java]: It will be "lazy" loaded even when created in static constructor. If no field in CustomNotification will never be accessed, then static constructor will never be called too.

**Comment by deathbeam** [comment]
This needs rebase. Also, after the recent notification system changes, maybe this notification should be defined in the NotificationMode now instead. And, make it default to the custom tray notifications on windows, otherwise to just tray notifications. Even though I still don't really agree with having the system-based default.

**Comment by Abextm** [comment]
Fixes #2275

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java [hunk 12]
```diff
@@ -73,51 +95,84 @@ default boolean enableCustomChrome()
...
 		return true;
 	}
 
+	@ConfigItem(
+		keyName = "notificationGameMessage",
+		name = "Enable game message notifications",
+		description = "Puts a notification message in the chatbox",
+		position = 23
+	)
+	default boolean enableGameMessageNotification()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: This should be reverted to previous behaviour, there is no reason to change it. The intention of the new notification types was to not use the together with other options, and it should stay that way. Also, this is completely out of scope of this PR?

File: runelite-client/src/main/java/net/runelite/client/Notifier.java [hunk 13]
```diff
@@ -336,4 +348,46 @@ private static String toUrgency(TrayIcon.MessageType type)
...
+		Clip clip = null;
+
+		// Try to load the user sound from ~/.runelite/notification.wav
+		File file = new File(RuneLite.RUNELITE_DIR, "notification.wav");
+		if (file.exists())
+		{
+			try
+			{
+				InputStream fileStream = new BufferedInputStream(new FileInputStream(file));
+				try (AudioInputStream sound = AudioSystem.getAudioInputStream(fileStream))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/Notifier.java]: I feel like the sound should be pre-loaded on config change, this will be running in game thread

File: runelite-client/src/main/java/net/runelite/client/Notifier.java [hunk 14]
```diff
@@ -336,4 +348,46 @@ private static String toUrgency(TrayIcon.MessageType type)
 				return "normal";
 		}
 	}
+
+	private void playCustomSound()
+	{
+		Clip clip = null;
+
+		// Try to load the user sound from ~/.runelite/notification.wav
+		File file = new File(RuneLite.RUNELITE_DIR, "notification.wav");
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/Notifier.java]: What about picking some more cross-platform format? .wav is not really ideal

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java (hunk 12 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: I changed it here because the behavior introduced in #2180 is broken. It does not work. It is in this PR, rather than a separate one because it would be impossible to write this one without depending on it, so there is no reason. If we want to merge 093b11f before this, I will cherry pick it onto a new branch and PR it separately, but I didn't think that was necessary.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Based on your use-case, you want only sound with tray notifications. What is also discutable because it looks like it is problem either in Java or in your OS/configuration. But, with custom sound it at least makes sense as separate option and I would do it like that as well. Otherwise I explained the reasoning for the modes, also I dislike pointless toggles, so that is why it is in dropdown and not each type of notification separate. That is also common use-case, so unless there is good reason to change it, it should not be changed.

File: runelite-client/src/main/java/net/runelite/client/Notifier.java (hunk 13 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/Notifier.java]: I've opted to just put this in the executor, because I don't want to have to deal with support for people's custom sounds not loading.

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java (hunk 12 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: My preferred configuration is [Custom tray/Custom sound/Chat message]. I see no reason other than "I dislike customization" for not having this separate.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Well as I said before, your preferred configuration is not useful. Tray notifications are useful when not looking at screen, chat notifications are useful when looking at screen. They server different purposes and they was added under intention to serve for that purposes.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: I use the chat messages as a way to gauge how much I'm paying attention, as they are moderately temporal.

**Comment by ShaunDreclin** [comment]
👀 I need this

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java (hunk 12 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Hmm. Well, I still disagree but I guess. But, this should be different PR on top of this one still so it can be decided there and not block this one.

File: runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java [hunk 15]
```diff
@@ -127,14 +127,16 @@ default boolean requestFocusOnNotification()
...
+		keyName = "notificationSound_",
+		name = "Notification sound",
 		description = "Enables the playing of a beep sound when notifications are displayed",
 		position = 22
 	)
-	default boolean enableNotificationSound()
+	default Notifier.NativeCustomOff notificationSound()
 	{
-		return true;
+		return System.getProperty("os.name").toLowerCase().contains("win")
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: This should be set to OFF by default imo.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: I set this to what most application default to: Toast+Sound.

I could add a OsSpecific<T> type that will marshal a map<OSName, T> in the config.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Most applications default to using system notification. What puzzles me is why you chose custom notification on Windows as default. With os-specific config you are probably taking your PR way too far beyond is original intention, what was just add custom notifications, but now you are also adding custom sounds, changing RuneLite settings, all in one PR, I don't understand why. The custom sounds at least make sort of sense, so be it. But, why custom sound should be default for Windows, that do not makes sense. But anyway, as long as the os-specific configuration is not resolved, I would avoid it. Also, there is already code for detecting OS, why are you not using that one.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: The reason these are both on by default on windows is that the windows-native notifications are garbage. Their clickbox is roughly double size of the UI element. Clicking on them does not make them go away. Disabling them doesn't remove the clickbox. They force applications out of full screen mode. You cannot turn the volume down without editing the file. They do not work over remote desktop. They are __entirely broken__. Many applications (Chrome, Firefox, Anything using electron) that I use have no user-facing option to use native notifications on windows.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Well, still better than having os-specific persisted configuration. And, the users have ability to choose. Also, what is the reasoning for custom sound default on windows instead of system beep, noone was reporting problems with that one.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Because the system beeps (and notifications) are handled in a weird way, you cannot easily turn the volume down, so if you have most applications turned down, they will deafen you. They are two halves of the same problem, so I figured I should fix them at the same time.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Also the system beep is not the notification sound, though they are the same by default on windows. Also people have complained about it. See #2275

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: That do not sounds like problem to me, as people are used to system beep on Windows, and I would assume they have tweaked the volume of it accordingly. Also it respects your volume setting of your OS, so it is non-issue. And, that is not complain, that is feature request for having the sound customizable, like you did, but still, no reason to make it default behaviour. Like, I understand that you have this problem, but, there is option to fix it, so why not leave the system-specific settings out of this.

Comment (ShaunDreclin) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Yeah the system beep is incredibly loud compared to every other audio source. I have to go into the volume mixer and reduce the volume for the app that sent the system beep to get it to a level that doesn't kill my ears

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Yeah that sounds like correct behaviour, I doubt any native OS notifications respect per-application sound, as notifications are not tied to applications, applications just send them to some handler. And yeah, well, you can change the config option, like everyone else after it will be added.

Comment (ShaunDreclin) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: That said, having different default configs based on OS is probably unnecessary, just providing the option is enough.

Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: I wouldn't mind defaulting it to off, if the option was visible in any way, but it is hidden.

Comment (ShaunDreclin) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: Maybe putting the `RuneLite` config section at the top of the menu would help with that.

Also, do we have any data on OS usage? If the majority of the players use windows then custom notifications should def be enabled by default

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/config/RuneLiteConfig.java]: With this I would result to by default not break anything for anyone. And custom notifications do mostly that for other OSes.

**Comment by LeviSchuck** [comment]
Dat gif
https://www.youtube.com/watch?v=jUy9_0M3bVk

**Comment by LeviSchuck** [comment]
That notification.wav sound is painful.

**Comment by deathbeam** [comment]
Use the enum for OS, also needs rebase.
